docs/source/contributing.rst

```diff
@@ -0,0 +1,172 @@
+
+Also see this page on `how to update your fork to the upstream repository`_.
+
+.. _GitHub's forks: https://guides.github.com/activities/forking/
+.. _how to update your fork to the upstream repository: https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a
+
+Development Environment
+=======================
+To set up the development environment on your computer, change to the local
+repository and:
+
+* Create the virtual environment in ``venv`` directory:
+
+  .. code-block::
+
+      python -m venv venv
+
+
+* Activate the virtual environment, on Windows:
+
+  .. code-block::
+
+      venv\Scripts\activate
+
+
+  or on Linux/Mac:
+
+  .. code-block::
+
+      . venv/bin/activate
+
+* Install the test dependencies and documentation dependencies of
+  ``asyncstdlib``:
+
+  .. code-block::
+
+      pip3 install .[test,doc]
+
+Write the Code
+==============
+
+Now you can implement your feature.
+
+Make sure you also write the corresponding unit tests in the
+``unittests/`` directory.
+
+Pre-commit Checks
+=================
+We perform the following battery of pre-commit checks:
+
+* Check the code style with `flake8`_:
+
+  .. code-block::
+
+      python -m flake8
+
+* Check the code formatting with `Black`_:
+
+  .. code-block::
+
+      python -m black --target-version py36 --diff --check asyncstdlib unittests
+
+  You can also automatically format the code with:
+
+  .. code-block::
+
+    python -m black --target-version py36 asyncstdlib unittests
+
+* Run unit tests and measure the code coverage by:
+
+  .. code-block::
+
+      python -m coverage run -m pytest
+
+  The HTML report of the coverage can be generated by:
+
+  .. code-block::
+
+      coverage html
+
+  The report on coverage is stored in the ``htmlcov/`` directory.
+
+  Make sure that your test provide a full coverage of your code.
+
+.. _flake8: https://flake8.pycqa.org/en/latest/
+.. _Black: https://github.com/psf/black
+
+Document
+========
+Once you are finished with the implementation, do not forget to document your
+code in the directory ``docs/source/api/``.
+
+We use `Sphinx`_ to render the documentation to HTML.
+
+.. _Sphinx: https://www.sphinx-doc.org/en/master/
+
+To generate the documentation, change to ``docs/`` and execute, on Windows:
+
+.. code-block::
+
+    make.bat html
+
+or on Linux/Mac:
+
+.. code-block::
+
+    make html
+
+The documentation is available in ``docs/_build/html``.
+
+Commit your Changes
+===================
+Make a single commit of your changes.
+
+Write the subject of the commit message in `imperative mood`_.
+The subject should not exceed 50 characters.
+
+The body of the commit message should be a bullet list of changes and additions,
+usually in `past tense`_.
+Please observe the maximum line length of 72 characters for the body.
+
+Here is an example of a commit message:
+
+.. code-block::
+
+    Add itertools.pairwise
+
+    * added itertools.pairwise
+
+    * documented itertools.pairwise
```

Comment:
Can you please change that? This is basically the opposite of what I prefer: many small commits, each with one change. There shouldn't actually be a need to have a body for a commit message – that's a sure sign the commit is too large.  
The singular commit + summary is created when merging the branch into master.

Quick sketch of a possible description:
```suggestion
Make a single commit for each self-contained change.

Write the subject of the commit message in `imperative mood`_.
The subject should not exceed 50 characters.

Here is an example of two commit messages:

.. code-block::

    added itertools.pairwise

    documented itertools.pairwise
```

Reply:
@maxfischer2781 the current commit history uses present tense for the subject, while you used past tense in your examples here. What should be official?

Reply:
Going through the recent commit history, the largest bulk uses *past tense* for the commit messages and *present tense* for PR messages. So, commits should be *past tense*.

Reply:
@maxfischer2781 thanks! I changed the text and pushed another commit.